Correct breaking changes from new version

Iddingsite opened the pull request:
With the last version of MAGEMin_C, the new syntax for `single_point_minimization()`and `multi_point_minimization()`was introduced to allow multiple dispatch on the compositions. 

I think it would be nice to keep those as deprecated functions as it is a breaking change and the published version of MAGEMin_C is not a breaking one (still 1.3.X).
 
So I've created two new functions `single_point_minim`and `multi_point_minim`that uses the new syntax and the previous ones now send a warning.

So now this code:

```
    data    = Initialize_MAGEMin("ig", verbose=false);

    # One bulk rock for all points
    P,T     = 10.0, 1100.0
    Xoxides = ["SiO2"; "Al2O3"; "CaO"; "MgO"; "FeO"; "Fe2O3"; "K2O"; "Na2O"; "TiO2"; "Cr2O3"; "H2O"];
    X       = [48.43; 15.19; 11.57; 10.13; 6.65; 1.64; 0.59; 1.87; 0.68; 0.0; 3.0];
    sys_in  = "wt"
    out     = single_point_minimization(P, T, data, X=X, Xoxides=Xoxides, sys_in=sys_in)
```

output

```
┌ Warning: `single_point_minimization(P,T,MAGEMin_db;test=0,X=nothing,B=nothing,Xoxides=Vector{String},sys_in="mol",progressbar=true)` is deprecated, use `single_point_minim(P,T,MAGEMin_db,X;test=test,B=B,Xoxides=Xoxides,sys_in="mol",progressbar=true)` instead.
```

before the printing.

I've also done a bit cleaning. If the tests pass and you are good with that, I will also propagate the changes to MAGEMin.

boriskaus commented:
Why don’t you keep the old name for the new function, but make a multiple dispatch (and a warning)? So we call both old & new `single_point_minimization`?

Iddingsite commented:
I don't think that works because they are in fact 2 different fonctions. One has X as a positional argument, the Other one has it as a keyword argument. If you give them the same name with one calling the other one, you will have a warning when you use using MAGEMin_C that you have 2 fonctions with the same name. Hence the New function name. 
Happy to be proven wrong!

boriskaus commented:
It’s called ‘multiple dispatch’ and is one of the reasons that julia is so powerful.

Iddingsite commented:
Yeah sure, but multiple dispatch only works with positional arguments, not keyword arguments. The old function takes X as keyword argument, Hence not allowing multiple dispatch on this variable. That's why I changed the old syntaxe and why they are not compatible for me

NicolasRiel commented:
Yes I tried it out, and I think Hugo is right, it does not work with the same arguments shuffled differently:

`Info Given MAGEMin_C was explicitly requested, output will be shown live 
WARNING: Method definition single_point_minimization(Float64, Float64, MAGEMin_C.MAGEMin_Data{TypeGV, TypeZB, TypeDB, TypeSplxData} where TypeSplxData where TypeDB where TypeZB where TypeGV) in module MAGEMin_C at /home/seph42/.julia/dev/MAGEMin_C/julia/MAGEMin_wrappers.jl:168 overwritten at /home/seph42/.julia/dev/MAGEMin_C/julia/MAGEMin_wrappers.jl:203.
ERROR: Method overwriting is not permitted during Module precompilation. Use `__precompile__(false)` to opt-out of precompilation.`

Still for now I would not change the public version of MAGEMin_C as it implies other changes.

Let's have it working correctly on the dev branch first

Iddingsite commented:
Ok, after discussion with @NicolasRiel, we think that changing the name is the way to go. The warning is also only running if julia is started with --depwarn=yes for now.

boriskaus commented:
I’d like to check - yet I will only get around doing this tomorrow

Iddingsite commented:
> I’d like to check - yet I will only get around doing this tomorrow

Fine by me. If you find a more elegant way, I will be interested. 

Also, still failing in 1.6, which was not the case before my last 2 push so I did somewhere there. I will stop working on it today, but I will have a look tomorrow.

ranocha commented:
Do I understand correctly that the breaking change is basically what was done in https://github.com/ComputationalThermodynamics/MAGEMin_C.jl/pull/21?

If so, you don't really perform dispatch here - you just restrict the types of some arguments. That can also be done for keyword arguments:
```julia
julia> function positional(X::T2) where {T2 <: VecOrMat}
           println("positional")
       end
positional (generic function with 1 method)

julia> function keyword(; X::VecOrMat = nothing)
           println("keyword")
       end
keyword (generic function with 1 method)

julia> let X = nothing
           positional(X)
           keyword(; X)
       end
positional
keyword

julia> let X = [1.0, 2.0]
           positional(X)
           keyword(; X)
       end
positional
keyword

julia> let X = range(0.0, 1.0, length = 3)
           positional(X)
           keyword(; X)
       end
positional
keyword

julia> let X = [range(0.0, 1.0, length = 3)]
           positional(X)
           keyword(; X)
       end
positional
keyword
```

Iddingsite commented:
That's great Hendrick, that would work 👍 I will go back to the old syntaxe with this approch. May still be interesting to have multiple dispatch on the composition at some point (e.g. to convert some specific array type in something that MAGEMin Can use) but for my applications I don't need it.

ranocha commented:
Even if you need multiple dispatch, you can handle that indirectly via keyword arguments:

```julia
julia> outer(; a) = inner(a)
outer (generic function with 1 method)

julia> inner(::AbstractFloat) = println("Float")
inner (generic function with 1 method)

julia> inner(::Integer) = println("Integer")
inner (generic function with 2 methods)

julia> outer(a = 1.0)
Float

julia> outer(a = 1)
Integer
```

Iddingsite commented:
Ok, yeah that's smart. Nice if we have a workaround!

Iddingsite commented:
Ok, the tests are failing because there is a mutable struct in this PR (changes made my @NicolasRiel to had access to Margules parameters for inversions), that's why it is failing in 1.6 and 1.7. I would say we could drop support for this versions or remove that. I don't know if the mutable is needed or not.

Iddingsite commented:
Ok, looks good to be merged to me. 

I've also modified P and T to AbstractVector (using multiple dispatch this time) for `multi_point_minimization`and added a test for `@view`arrays for these 2. It doesn't change anything to existing code for me as it is a minor change and not worth a different PR but tell me if you think otherwise.

NicolasRiel commented:
> Ok, the tests are failing because there is a mutable struct in this PR (changes made by @NicolasRiel to have access to Margules parameters for inversions), that's why it is failing in 1.6 and 1.7. I would say we could drop support for these versions or remove that. I don't know if the mutable is needed or not.

This is not the only the only mutable structure. And actually the error points to the once that was already there. Something is fishy

Iddingsite commented:
error was pointing at line 26 which is the beginning of struct W_data and its docstring. I just removed the {T2,T2,M1}
Check that it is still doing what you want, because I did not write that part.
